website/payments/admin.py

```diff
@@ -285,6 +387,126 @@ def queryset(self, request, queryset) -> QuerySet:
+    list_display = (
+        "description",
+        "start_date",
+        "end_date",
+        "total_amount",
+        "payments_count",
+        "processing_date",
+        "processed",
+    )
+    fields = (
+        "description",
+        "processed",
+        "processing_date",
+        "total_amount",
+    )
+    search_fields = ("description",)
+
+    def get_readonly_fields(self, request: HttpRequest, obj: Batch = None):
+        default_fields = (
+            "processed",
+            "processing_date",
+            "total_amount",
+        )
+        if obj and obj.processed:
+            return ("description",) + default_fields
+        return default_fields
+
+    def delete_model(self, request, batch):
+        if batch.processed:
+            messages.error(request, "You cannot delete a processed batch.")
+        else:
+            super().delete_model(request, batch)
+
+    def has_delete_permission(self, request, obj=None):
+        if isinstance(obj, Batch):
+            if obj.processed:
+                return False
+
+        return super().has_delete_permission(request, obj)
+
+    def get_urls(self) -> list:
+        urls = super().get_urls()
+        custom_urls = [
+            path(
+                "<uuid:pk>/process/",
+                self.admin_site.admin_view(admin_views.BatchProcessAdminView.as_view()),
+                name="payments_batch_process",
+            ),
+            path(
+                "<uuid:pk>/export/",
+                self.admin_site.admin_view(admin_views.BatchExportAdminView.as_view()),
+                name="payments_batch_export",
+            ),
+            path(
+                "new_filled/",
+                self.admin_site.admin_view(
+                    admin_views.BatchNewFilledAdminView.as_view()
+                ),
+                name="payments_batch_new_batch_filled",
+            ),
+        ]
+        return custom_urls + urls
+
+    def save_formset(self, request, form, formset, change):
+        instances = formset.save(commit=False)
+
+        for instance in instances:
+            if instance.batch is None or (
+                instance.batch and not instance.batch.processed
+            ):
+                instance.batch = None
+            instance.save()
+        formset.save_m2m()
+
+    def changeform_view(
+        self,
+        request: HttpRequest,
+        object_id: str = None,
+        form_url: str = "",
+        extra_context: dict = None,
+    ) -> HttpResponse:
+        """
+        Renders the change formview
+        Only allow when the batch has not been processed yet
+        """
+        extra_context = extra_context or {}
+        obj = None
+        processed = False
+        if object_id is not None and request.user.has_perm("payments.process_batches"):
+            obj = Batch.objects.get(id=object_id)
+            processed = obj.processed
+
+        extra_context["batch"] = obj
+        extra_context["processed"] = processed
```

Comment:
```suggestion
        obj = None
        if object_id is not None and request.user.has_perm("payments.process_batches"):
            obj = Batch.objects.get(id=object_id)

        extra_context["batch"] = obj
```
You don't need to add this processed, you already have the batch and can access it in the template like `batch.processed`.